src/Standards/Squiz/Sniffs/WhiteSpace/OperatorSpacingSniff.php

```diff
@@ -325,6 +325,12 @@ public function process(File $phpcsFile, $stackPtr)
      */
     protected function isOperator(File $phpcsFile, $stackPtr)
     {
+        $rightHandSide = $phpcsFile->findNext(Tokens::$emptyTokens, ($stackPtr + 1), null, true);
```

Comment:
Why are you using `Tokens::$emptyTokens` instead of `T_WHITESPACE` here ?

Reply:
Because `<?php $a = /* one */ 1;` is a valid assignment, and `<?php $a = /* one */` is a parse error.

This highlights another test case which should probably be added: `<?php $a = ;` Would you like me to add this test in scope of this pull request, or is that something for another time? I've checked just now and this branch / pull request already handles these scenarios.

Reply:
> Because `<?php $a = /* one */ 1;` is a valid assignment, and `<?php $a = /* one */` is a parse error.

True, but that's irrelevant for these sniffs. In both cases, the sniffs check that there is one space between the operator and the comment and it can do that check, even if there is nothing after the comment.

Then again... if there is a parse error, the "skip default values in function declarations + skip declare statements" check at the start of the `isOperator()` method is unreliable, which is really the heart of the problem which this PR is trying to solve.

This also basically confirms that the parse error check needed to be moved up.

Looking at the sniffs again - the sniffs already contain defensive coding against parse errors and realistically that defensive coding is sufficient, except for the "skip default values in function declarations + skip declare statements" check - which is limited to only apply to two operator tokens, while the sniff checks for a lot more tokens.

So maybe the new defensive coding being added should only be applied to the "skip default values in function declarations + skip declare statements" check ? in which case, using `Tokens::$emptyTokens` instead of `T_WHITESPACE` would be correct.

@fredden What do you think ?


> This highlights another test case which should probably be added: `<?php $a = ;` Would you like me to add this test in scope of this pull request, or is that something for another time? I've checked just now and this branch / pull request already handles these scenarios.

I think we should leave that for another PR and probably check for fixer conflicts on that code snippet with sniffs which check the spacing before a semi-colon, like the `Squiz.WhiteSpace.SemicolonSpacing` sniff.

Reply:
I've looked through <https://www.php.net/manual/en/language.operators.php>. Nearly all operators require something to their right in order to be valid PHP. The exceptions here are the post-increment and post-decrement operators (`$a++` and `$a--`), and the second tick operator (which closes the first). However, in those cases, we still need a non-comment token (eg `;` or `?>`) in order to not be a parse error.

With this in mind, I think it's fair to say that a particular token (which might be an operator) is part of a parse error if it's not followed by a non-comment token. I'd like to keep the parse-error detection as-is, and at the top of the method.

I tried moving the check inside the if-block that you mentioned, and the current tests in this pull request still pass. However the parse error `<?php $a+/* comment */` is no longer detected as a parse error and the sniff raises errors. Shall I add this as a test case here too?

Reply:
@fredden The thing is - the sniff is basically split in two parts:
1. Check spacing _before_ the operator.
2. Check spacing _after_ the operator.

The "is operator?" check is intended to check if the sniff needs to run at all.

The difference between what you are proposing and what I'm suggesting is this:
* With the parse error check based on "code after" at the start of the "is operator?" check, the sniff bows out early for a parse error and won't run at all during live coding.
* With the parse error check limited to the "declare statement/function declaration" check within the "is operator?" check, the sniff would still run the "spacing _before_ the operator" part during live coding.
    The sniff `process()` method already contains defensive coding to bow out between the _before_ and _after_ check if there is nothing after.

> However the parse error <?php $a+/* comment */ is no longer detected as a parse error and the sniff raises errors. 

And so it should.
The spacing _before_ the `+` can be checked and the spacing _after_ can be checked as well (as there is a comment).

Try a code sample like this: `<?php $a+`.
Here you will see that the sniff checks only the spacing before, but not the spacing after (with the original code) and that is fine as the spacing before the operator can be checked already, while the spacing after cannot be checked due to the parse error.
With the change as you are proposing now, that snippet would be ignored completely in live coding until the statement would be complete.

Does that make it clearer why I'm suggesting to limit the additional defensive coding to the "declare statement/function declaration" check within the "is operator?" check ?

Reply:
We discussed this further off-GitHub and decided on a way forward to get this in a merge-ready state.